pylib/cassandra-cqlsh-tests.sh

```diff
@@ -23,79 +23,83 @@
 
-if [ -z "$CASSANDRA_USE_JDK11" ]; then
-    export CASSANDRA_USE_JDK11=false
+if [[ -z "${version}" ]]; then
+  version=$(grep 'property\s*name=\"base.version\"' ${CASSANDRA_DIR}/build.xml |sed -ne 's/.*value=\"\([^"]*\)\".*/\1/p')
 fi
 
-if [ "$CASSANDRA_USE_JDK11" = true ] ; then
-    TESTSUITE_NAME="${TESTSUITE_NAME}.jdk11"
-else
-    TESTSUITE_NAME="${TESTSUITE_NAME}.jdk8"
-    unset JAVA11_HOME
+if [[ -z "${java_version}" ]]; then
+  java_version=$(java -version 2>&1 | awk -F '"' '/version/ {print $2}' | sed 's/^1\.//g' | awk -F. '{print $1}')
+fi
+if [[ ${java_version} -ge 11 ]]; then
+  export CASSANDRA_USE_JDK11=true
 fi
 
-ant -buildfile ${CASSANDRA_DIR}/build.xml realclean
-# Loop to prevent failure due to maven-ant-tasks not downloading a jar..
-for x in $(seq 1 3); do
-    ant -buildfile ${CASSANDRA_DIR}/build.xml jar
-    RETURN="$?"
-    if [ "${RETURN}" -eq "0" ]; then
-        break
-    fi
-done
-# Exit, if we didn't build successfully
-if [ "${RETURN}" -ne "0" ]; then
-    echo "Build failed with exit code: ${RETURN}"
-    exit ${RETURN}
+if [[ -z "${python_version}" ]]; then
+  python_version="3.8"
+  command -v python3 >/dev/null 2>&1 && python_version="$(python3 -V | awk '{print $2}' | awk -F'.' '{print $1"."$2}')"
 fi
 
+export TESTSUITE_NAME="cqlshlib.python${python_version}.jdk${java_version}"
+
+pushd ${CASSANDRA_DIR} >/dev/null
+
+# check project is already built. no cleaning is done, so jenkins unstash works, beware.
+[[ -f "${BUILD_DIR}/dse-db-${version}.jar" ]] || [[ -f "${BUILD_DIR}/dse-db-${version}-SNAPSHOT.jar" ]] || { echo "Project must be built first. Use \`ant jar\`. Build directory is ${BUILD_DIR} with: $(ls ${BUILD_DIR})"; exit 1; }
+
 # Set up venv with dtest dependencies
 set -e # enable immediate exit if venv setup fails
-virtualenv --python=$PYTHON_VERSION venv
-source venv/bin/activate
-# 3.11 needs the newest pip
-curl -sS https://bootstrap.pypa.io/get-pip.py | $PYTHON_VERSION
 
-pip install -r ${CASSANDRA_DIR}/pylib/requirements.txt
-pip freeze
+# fresh virtualenv and test logs results everytime
+rm -fr ${DIST_DIR}/venv ${DIST_DIR}/test/{html,output,logs}
+
+if [ "$cython" = "yes" ]; then
+  cython_suffix="cython"
+else
+  cython_suffix="no-cython"
+fi
+
+# re-use when possible the pre-installed virtualenv found in the cassandra-ubuntu2004_test docker image
+virtualenv-clone ${BUILD_HOME}/${cython_suffix}/python${python_version} ${BUILD_DIR}/venv || virtualenv --python=python${python_version} ${BUILD_DIR}/venv
+source ${BUILD_DIR}/venv/bin/activate
 
 if [ "$cython" = "yes" ]; then
-    TESTSUITE_NAME="${TESTSUITE_NAME}.cython"
-    pip install "Cython>=0.29.15,<3.0"
-    cd pylib/; python setup.py build_ext --inplace
-    cd ${WORKSPACE}
+  pip install "Cython>=0.29.15,<3.0"
+  unset CASS_DRIVER_NO_CYTHON
 else
-    TESTSUITE_NAME="${TESTSUITE_NAME}.no_cython"
+  export CASS_DRIVER_NO_CYTHON=true
+fi
+
+pip install --exists-action s -r ${CASSANDRA_DIR}/pylib/requirements.txt
```

Comment:
```suggestion
pip install --exists-action w -r ${CASSANDRA_DIR}/pylib/requirements.txt
```
That is how we do it in Apache and makes sense to me to not rely on caching things

Reply:
Updates in those repos are quite rare. On the other hand, cython compilation of Cassandra driver takes quite a long time and we would have to do that every time we run a PR build, which means quite often. I don't want us to waste that time. The docker image is automatically rebuilt and republished every week so we will get updates of those repos at that frequency

Reply:
Makes sense
I did not know this:
`The docker image is automatically rebuilt and republished every week so we will get updates of those repos at that frequency`

In OSS it is updated once a year or so....